services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/AbstractPushMonitor.java

```diff
@@ -927,35 +938,77 @@ protected void checkWhetherToStartBufferReplayForHybrid(OfflinePushStatus offlin
-        } catch (Exception e) {
-          // TODO: Figure out a better error handling...
-          String newStatusDetails = "Failed to kick off the buffer replay";
-          handleTerminalOfflinePushUpdate(offlinePushStatus, new ExecutionStatusWithDetails(ERROR, newStatusDetails));
-          LOGGER.error("{} for offlinePushStatus: {}", newStatusDetails, offlinePushStatus, e);
+          newStatusDetails.append("kicked off buffer replay");
+        } else if (!offlinePushStatus.getCurrentStatus().isTerminal()) {
+          LOGGER.info(
+              "{} is not ready to start buffer replay. Current state: {}",
+              offlinePushStatus.getKafkaTopic(),
+              offlinePushStatus.getCurrentStatus().toString());
         }
-      } else if (!offlinePushStatus.getCurrentStatus().isTerminal()) {
-        LOGGER.info(
-            "{} is not ready to start buffer replay. Current state: {}",
-            offlinePushStatus.getKafkaTopic(),
-            offlinePushStatus.getCurrentStatus().toString());
       }
+
+      if (isEOPReceivedInAllPartitions) {
+        // Check whether to send EOP for materialized view topic(s)
+        for (ViewConfig rawView: viewConfigMap.values()) {
+          VeniceView veniceView = ViewUtils.getVeniceView(
+              rawView.getViewClassName(),
+              new Properties(),
+              store.getName(),
+              rawView.getViewParameters());
+          if (veniceView instanceof MaterializedView) {
+            MaterializedView materializedView = (MaterializedView) veniceView;
+            for (String materializedViewTopicName: materializedView
+                .getTopicNamesAndConfigsForVersion(version.getNumber())
+                .keySet()) {
+              VeniceWriterOptions.Builder vwOptionsBuilder =
+                  new VeniceWriterOptions.Builder(materializedViewTopicName).setUseKafkaKeySerializer(true)
+                      .setPartitionCount(materializedView.getViewPartitionCount());
+              try (VeniceWriter veniceWriter = veniceWriterFactory.createVeniceWriter(vwOptionsBuilder.build())) {
+                veniceWriter.broadcastEndOfPush(Collections.emptyMap());
+              }
+              if (newStatusDetails.length() > 0) {
+                newStatusDetails.append(", ");
+              }
+              newStatusDetails.append("broadcast EOP to materialized view topic: ").append(materializedViewTopicName);
+            }
+          }
+        }
+        updatePushStatus(
+            offlinePushStatus,
+            ExecutionStatus.END_OF_PUSH_RECEIVED,
+            Optional.of(newStatusDetails.toString()));
+        LOGGER.info("Successfully {} for offlinePushStatus: {}", newStatusDetails.toString(), offlinePushStatus);
+      }
+    } catch (Exception e) {
+      String newStatusDetails = "Failed to start EOP procedures";
+      handleTerminalOfflinePushUpdate(offlinePushStatus, new ExecutionStatusWithDetails(ERROR, newStatusDetails));
```

Comment:
Controller has logics to mutate version topics in the following scenarios:
1. Version topic will be deleted when the version is being deleted and Controller should delete the corresponding view topics as well.
2. Controller will enable log compaction in version topic for hybrid store when push job is completed, and I think we should enable it for view topics as well and in theory. 

Do we have logic to cover the above cases?

Reply:
1. is already handled today via `VeniceHelixAdmin#cleanUpViewResources`.
2. Good call, I've added the logic to also update log compaction for hybrid store's view topics.